crates/typst-library/src/text/quote.rs

```diff
@@ -0,0 +1,136 @@
+    ///
+    /// - `{true}`: Wrap this quote in double quotes.
+    /// - `{false}`: Do not wrap this quote in double quotes.
+    /// - `{auto}`: Infer whether to wrap this quote in double quotes based on
+    ///   the `block` property. If `block` is `{false}` double quotes are
+    ///   auomatically added.
+    ///
+    /// ```example
+    /// #set text(lang: "de")
+    /// #quote[Ich bin ein Berliner.]
+    ///
+    /// #set smartquote(quotes: "«»")
+    /// #set quote(quotes: true)
+    /// #quote(block: true)[
+    ///   ... ἔοικα γοῦν τούτου γε σμικρῷ τινι αὐτῷ τούτῳ σοφώτερος εἶναι, ὅτι
+    ///   ἃ μὴ οἶδα οὐδὲ οἴομαι εἰδέναι.
+    /// ]
+    /// ```
+    quotes: Smart<bool>,
+
+    /// The source url to this quote.
+    ///
+    /// ```example
+    /// #show link: set text(blue)
+    /// #quote(source: "https://google.com")[cogito, ergo sum]
+    /// ```
+    source: Option<EcoString>,
+
+    /// The author of this quote.
+    ///
+    /// ```example
+    /// #quote(author: [René Descartes])[cogito, ergo sum] is the author. \
+    /// #quote[cogito, ergo sum] --- _Unknown_
+    /// ```
+    author: Option<Content>,
+
+    /// The quote.
+    #[required]
+    body: Content,
+}
+
+impl Show for QuoteElem {
+    fn show(&self, _: &mut Vt, styles: StyleChain) -> SourceResult<Content> {
+        let mut realized = self.body();
+        let author = self.author(styles);
+        let block = self.block(styles);
+
+        if self.quotes(styles).as_custom().is_some_and(|q| q) || !block {
+            let quote = SmartquoteElem::new().with_double(true).pack();
+            let weak_h = HElem::new(Spacing::Rel(Rel::zero())).with_weak(true).pack();
+
+            realized = Content::sequence([
+                quote.clone(),
+                weak_h.clone(),
+                realized,
+                weak_h,
+                quote,
+            ]);
+        }
+
+        if block {
+            realized = BlockElem::new().with_body(Some(realized)).pack();
+
+            if let Some(author) = author {
+                realized += Content::sequence([
+                    TextElem::packed('—'),
+                    SpaceElem::new().pack(),
+                    author,
+                ])
+                .aligned(Align::END);
+            }
+
+            let pad: Rel = Em::new(1.0).into();
+            realized = PadElem::new(realized).with_left(pad).with_right(pad).pack();
+        } else if let Some(author) = author {
+            let inline = |first: Content, second: Content| {
+                Content::sequence([
+                    first,
+                    SpaceElem::new().pack(),
+                    TextElem::packed('—'),
+                    SpaceElem::new().pack(),
+                    second,
+                ])
+            };
+
+            if TextElem::dir_in(styles) == Dir::LTR {
+                realized = inline(realized, author);
+            } else {
+                realized = inline(author, realized);
+            }
```

Comment:
Do we really need explicit code for the swapping? Without it, an English quote would be LTR in an RTL context, but an Arabic quote wouldn't be. The swapping already happens implicitly because the content ends up right to left. It just isn't obversable for English text because BiDi is smart enough not to write English in RTL.

Reply:
I'm not quite sure when exactly content is rearranged for RTL, I noticed that in RTL:
- `[a --- b]` stays `[a --- b]`
- `[a ---]` becomes `[--- a]`

I asked a colleague of mine from Jordan if inline quotes would follow my intuition of being `[author --- "quote"]` in say arabic.

So I want to display `[#quote --- #author]` for LTR and `[#author --- #quote]` for RTL, this doesn't happen automatically. For the block version, no swapping is needed because the em dash is leading.

But maybe it can be done better, I'm up for suggestions.

Reply:
`a --- b` stays `a --- b` because `a` and `b` are Latin letters and Latin is not an RTL script. Setting RTL doesn't force Latin to be written from right-to-left, it only sets the top-level text direction (which affects punctuation), but embedded LTR content stays RTL. If you try the same thing with actual RTL text, it is swapped:

 ```typst
#set text(lang: "ar", font: "Noto Sans Arabic")

// Each word in its own line so that the order is clear.
مرحبًا 
---
عالم

```
Result:
<img width="150" alt="image" src="https://github.com/typst/typst/assets/17899797/da474eba-34f8-4f45-b870-3c1ea4d5b21f">